### PR TITLE
Test branch name verification 199

### LINKED ISSUE #199: Unit tests for the branch verification
There should be unit tests to ensure that the branch verification works. They can also be used for regression testing.

- [x] Port existing bash script to python
- [ ] Implement unit tests

### PR DESCRIPTION
**Pull request by NinjaWeeaboo**
# Assignee Section

## Description
I ported the branch name verification script from bash to Python. In order to allow unit testing without the use of a new framework. I then wrote a testing script for Pytest, that tests each function within the name verification script. In order to be able to suppress the console outputs from `run_shell` so that it wouldn't print to the console on every checkout. I added a third argument that allows its print statements to be 'silenced'. By default it will still run its print statement without the third argument.

- [x] Rewrite branch verification script in Python to make unit testing easier.
- [x] Add testing script.
- [x] Add new 'silent' argument to `run_shell`.

closes #199 

Approval from at least one software team lead is necessary before merging.

# Reviewer Section

Aside from local testing and the General Integration Test it is implied that static analysis should be included in the verification process.

- [x] Local Test Performed Successfully
~~- [ ] [General Integration Test](https://docs.google.com/document/d/1ug0CpA1cIzURP8DDFSvCt2CEJJSwJ6Ta6B1LG_hYk6I/edit) Performed Successfully~~

For Pull Requests that do not include code changes, it is not required to perform the tests above.

# Testing guide
**Testing ported branch name verification script**
Due to the changes made to app.py, these steps will have to be followed in order to test the name verification script with `git checkout`. If they're not followed, errors related to the `run_shell` function will occur.
- Create new branch with name of choice.
- Merge test-branch-name-verification-199 into the new branch.
- Checkout between the two branches to confirm that the error message prints when expected to.

For the testing script, just run Pytest. I've confirmed that it works properly on Travis CI.

Bring on the roasts.

### CODE REVIEW

**Comment by NinjaWeeaboo** [comment]
I just realised that `run_shell` got moved to another python file in new-camera-template-201. Should I just make those same changes on this pr? Or should this just be dealt with after either one of these branches are merged into master?

File: branch_name_verification.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    if not is_excluded_branch(branch_name, excluded_branches):
+        errors[0] = has_upper_case_letters(trim_issue_num)
+        errors[1] = not has_hyphen_seperator(trim_issue_num)
+        errors[2] = not has_issue_num(branch_name)
+        errors[3] = has_invalid_symbols(branch_name)
+
+        if not all(errors[x] == False for x in errors):
+            print(error_msg)
+
+    return not all(errors[x] == False for x in errors)
```
Comment (MartensCedric) [inline branch_name_verification.py]: interesting alternative to an if statement.

To keep it clean I would recommend doing this instead.

```
is_name_valid = not has_upper_case_letters(trim_issue_num) and \
                  has_hyphen_separator(trim_issue_num) and \ 
                  ..........

```

then 
```
if not is_name_valid:
print(error_msg)
```

then you just `return is_name_valid`

Although, something else might be better. Do you have any suggestions @PeterGhimself ?

File: branch_name_verification.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    """
+    check for hyphen seperator between branch name and issue number
+    """
+    return trim_issue_num.endswith('-')
+
+def has_issue_num(branch_name):
+    """
+    check if there is an issue number
+    """
+    return not re.search(r'\d+$', branch_name) is None
```
Comment (MartensCedric) [inline branch_name_verification.py]: I recommend making this function retrieve the issue number and return it, if it is not there return None.

This will allow you to make unit tests to verify that it actually returns the correct number which at the moment you do not test.

If you really wanna keep this function then you can just do 

```
def has_issue_num(branch_name):
  return get_issue_num(branch_name) != None

File: robot/basestation/app.py [hunk 3]
```diff
@@ -22,20 +22,20 @@
 
 app = flask.Flask(__name__)
 
-def run_shell(cmd, args=""):
+def run_shell(cmd, args="", print_output=True):
```
Comment (MartensCedric) [inline robot/basestation/app.py]: Did you ever use the parameter you added?

File: branch_name_verification.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/env python
+import robot.basestation.app as app
+from robot.basestation.app import run_shell
+import re
+
+excluded_branches = {'master', 'develop', 'staging', 'test'}
+
+output, error = run_shell("git symbolic-ref --short HEAD", "", False)
```
Comment (MartensCedric) [inline branch_name_verification.py]: this function seems to be supposed to be used like this

`run_shell("git", "symbolic-ref --short HEAD", False)`

File: branch_name_verification.py (hunk 1 above)
Comment (MartensCedric) [reply branch_name_verification.py]: This is maybe better

```
  if cond1 == 'val1'\
     and cond2 == 'val2'\
     and cond3 == 'val3'\
     and cond4 == 'val4':
      do_something
```

I mean python is not good for this because of the way it doesnt work when stuff isnt formatted like it likes. But this kind of comes down to personal preferences.

Comment (MartensCedric) [reply branch_name_verification.py]: At this point it is almost personal preferences but in general I would advise not packing it into a list because you are evaluating all the conditions. (This means that all the functions are being executed). In the case where you have an many `and` statements, the first `false` will stop the executing of everything else in the condition thus being more optimal in terms of performance.

File: tests/unit/branch_name_test.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    """
+    test branch name checking function
+    """
+    assert branch_name_error('test-1', excluded_branches) == False
+    assert branch_name_error('Feat-test-branch-123', excluded_branches) == True
+    assert branch_name_error('feat-test-branch123', excluded_branches) == True
+    assert branch_name_error('feat-test-branch-', excluded_branches) == True
+    assert branch_name_error('fe@t-test-branch-123', excluded_branches) == True
+    assert branch_name_error('feat-test-branch-123', excluded_branches) == False
+    assert branch_name_error('upgrade-ubuntu-18-304', excluded_branches) == False
```
[the remark this answers is not in the document]
Comment (NinjaWeeaboo) [reply tests/unit/branch_name_test.py]: Good idea, I added a new test function to account for this

File: branch_name_verification.py (hunk 1 above)
Comment (NinjaWeeaboo) [reply branch_name_verification.py]: That's a lot cleaner than what I had done lol. I went with your first idea, thanks

File: robot/basestation/app.py (hunk 3 above)
Comment (NinjaWeeaboo) [reply robot/basestation/app.py]: I did, I used it in the branch name verification hook. In order to stop run_shell from printing stuff in the terminal after every checkout.

File: branch_name_verification.py (hunk 4 above)
Comment (NinjaWeeaboo) [reply branch_name_verification.py]: Fixed

File: branch_name_verification.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    return not get_issue_num(branch_name) is None
+
+def has_invalid_symbols(branch_name):
+    """
+    check for invalid invalid_symbols
+    """
+    invalid_symbols = {"/", "'", '!', '?', '@', '$', '%', ',', '.', '_', '+', '£', '=', '¬', '<', '>'}
+    return (bool(set(branch_name) & invalid_symbols)) or ('\\' in branch_name)
+
+def branch_name_error(branch_name, excluded_branches):
```
Comment (MartensCedric) [inline branch_name_verification.py]: since it returns a boolean I would call it has_branch_name_error

Comment (MartensCedric) [inline branch_name_verification.py]: I would not pass or check if the branch should be included in this function. It shouldn't be this functions responsibility to do both checking if it has an error and checking if it was excluded.

Comment (MartensCedric) [reply branch_name_verification.py]: I have a suggestion for what to do in another comment

File: branch_name_verification.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        and has_issue_num(branch_name)\
+        and not has_invalid_symbols(branch_name)
+
+        if not is_name_valid:
+            print(error_msg)
+
+    return not is_name_valid
+
+if __name__ == "__main__":
+    branch_name_error(branch_name, excluded_branches)
```
Comment (MartensCedric) [inline branch_name_verification.py]: if the branch is excluded, I recommend just not passing it to check if it has an error.

File: tests/unit/branch_name_test.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    assert is_excluded_branch('!@dvsfosk*&*&^fs4943', excluded_branches) == True
+
+def test_upper_case_letters():
+    """
+    test upper case letters checking
+    """
+    assert has_upper_case_letters('feat-test-branch-') == False
+    assert has_upper_case_letters('Feat-test-branch-') == True
+    assert has_upper_case_letters('feat-test-Branch-') == True
+    assert has_upper_case_letters('FEAT-TEST-BRANCH-') == True
```
Comment (MartensCedric) [inline tests/unit/branch_name_test.py]: should be removed

File: branch_name_verification.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
+#!/usr/bin/env python
+import robot.basestation.app as app
+from robot.basestation.app import run_shell
+import re
+
+excluded_branches = {'master', 'develop', 'staging', 'test'}
+
+output, error = run_shell("git", "symbolic-ref --short HEAD", False)
+branch_name = output.decode("utf8")[0:-1]
```
[the remark this answers is not in the document]
Comment (NinjaWeeaboo) [reply branch_name_verification.py]: Fixed

File: branch_name_verification.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    """
+    check if branch name is in list of excluded branches
+    """
+    return branch_name in excluded_branches
+
+def has_upper_case_letters(trim_issue_num):
+    """
+    check if trimed branch name only contains lower case characters
+    """
+    return not (trim_issue_num.islower())
```
[the remark this answers is not in the document]
Comment (NinjaWeeaboo) [reply branch_name_verification.py]: Shoot... well I removed it lol

File: branch_name_verification.py [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+def has_hyphen_seperator(trim_issue_num):
+    """
+    check for hyphen seperator between branch name and issue number
+    """
+    return trim_issue_num.endswith('-')
```
[the remark this answers is not in the document]
Comment (NinjaWeeaboo) [reply branch_name_verification.py]: I decided to give this the same treatment as has_upper_case_letters, and just removed it all together. I felt that having this function was useless, because `endswith()` already works on its own

File: branch_name_verification.py (hunk 6 above)
Comment (NinjaWeeaboo) [reply branch_name_verification.py]: Fixed

Comment (NinjaWeeaboo) [reply branch_name_verification.py]: Moved responsibility to where `__name__ == "__main__"` takes place

**Comment by NinjaWeeaboo** [comment]
`----------------------18` is comes back as invalid, because `islower()` returns false. Also to note, I have to use `islower()` because it and `isupper()` only return true if all of the letters in the string are lowercase or uppercase respectively. So I just use `not islower()` to check if there are any uppercase letters

File: tests/unit/branch_name_test.py [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+
+def test_excluded_branches():
+    """
+    test excluded branches checking
+    """
+    assert is_excluded_branch('test', excluded_branches) == False
+    assert is_excluded_branch('test-1', excluded_branches) == True
+    assert is_excluded_branch('!@dvsfosk*&*&^fs4943', excluded_branches) == True
+
+def test_get_issue_num():
```
Comment (PeterGhimself) [inline tests/unit/branch_name_test.py]: missing docstring

File: README.md [hunk 13]
```diff
@@ -159,7 +159,7 @@ This explains how to setup git hooks which prepend an issue number to a commit m
 Git hooks are important for performing repository validity checks. To setup git hooks using Git Bash, run the following commands from the root of the repository (ex: from ~/Programming/robotics-prototype/):
 
 - `cp commit-message-hook.sh .git/hooks/prepare-commit-msg`
-- `cp branch-name-verification.sh .git/hooks/post-checkout`
+- `cp branch_name_verification.py .git/hooks/post-checkout`
```
[the remark this answers is not in the document]
Comment (MartensCedric) [reply README.md]: agree

File: branch_name_verification.py [hunk 14]
```diff
@@ -0,0 +1,64 @@
+#!/usr/bin/env python
```
[the remark this answers is not in the document]
Comment (MartensCedric) [reply branch_name_verification.py]: agree

File: README.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (NinjaWeeaboo) [reply README.md]: Fixed

File: branch_name_verification.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (NinjaWeeaboo) [reply branch_name_verification.py]: Fixed

File: branch_name_verification.py [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+import robot.basestation.app as app
+from robot.basestation.app import run_shell
+import re
+
+excluded_branches = {'master', 'develop', 'staging', 'test'}
+
+output, error = run_shell("git", "symbolic-ref --short HEAD", False)
```
[the remark this answers is not in the document]
Comment (NinjaWeeaboo) [reply branch_name_verification.py]: Fixed

File: branch_name_verification.py [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+
+def is_excluded_branch(branch_name, excluded_branches):
+    """
+    check if branch name is in list of excluded branches
+    """
+    return branch_name in excluded_branches
+
+def get_issue_num(branch_name):
+    """
+    Gets issue number from branch named
```
[the remark this answers is not in the document]
Comment (NinjaWeeaboo) [reply branch_name_verification.py]: Fixed

File: branch_name_verification.py [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+    issue_num = re.search(r'\d+$', branch_name)
+    if issue_num is not None:
+        issue_num = issue_num.group(0)
+    return issue_num
+
+def has_issue_num(branch_name):
+    """
+    check if there is an issue number
+    """
+    return not get_issue_num(branch_name) is None
```
[the remark this answers is not in the document]
Comment (NinjaWeeaboo) [reply branch_name_verification.py]: Added more detail to docstrings

File: branch_name_verification.py [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+
+def has_invalid_symbols(branch_name):
+    """
+    check for invalid invalid_symbols
+    """
+    invalid_symbols = {"/", "'", '!', '?', '@', '$', '%', ',', '.', '_', '+', '£', '=', '¬', '<', '>'}
+    return (bool(set(branch_name) & invalid_symbols)) or ('\\' in branch_name)
```
[the remark this answers is not in the document]
Comment (NinjaWeeaboo) [reply branch_name_verification.py]: Added comments to explain this mess of a line

File: branch_name_verification.py [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+def has_double_hyphen(branch_name):
+    """
+    Checkout for double hyphens
+    """
+    return branch_name.find('--') != -1
+
+def has_branch_name_error(branch_name):
+    """
+    run all name verification functions and print error message if any fails
+    """
```
[the remark this answers is not in the document]
Comment (NinjaWeeaboo) [reply branch_name_verification.py]: Added

File: branch_name_verification.py [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+    trim_issue_num = re.sub(r'\d+$', '', branch_name)
+    error_msg = 'Branch name is not named properly, please see wiki for formating: https://github.com/space-concordia-robotics/robotics-prototype/wiki/Git-Workflow-and-Conventions'
```
[the remark this answers is not in the document]
Comment (NinjaWeeaboo) [reply branch_name_verification.py]: Fixed

File: branch_name_verification.py [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+    and has_issue_num(branch_name)\
+    and not has_invalid_symbols(branch_name)\
+    and not has_double_hyphen(branch_name)
+
+    if not is_name_valid:
+        print(error_msg)
+
+    return not is_name_valid
+
+if __name__ == "__main__" and not is_excluded_branch(branch_name, excluded_branches):
```
[the remark this answers is not in the document]
Comment (NinjaWeeaboo) [reply branch_name_verification.py]: Fixed, added a new if statement within `if __name__ == '__main__':` to handle checking if the branch is excluded

File: tests/unit/branch_name_test.py [hunk 22]
```diff
@@ -0,0 +1,60 @@
+import branch_name_verification as branch_name_verification
+from branch_name_verification import is_excluded_branch, get_issue_num, has_issue_num, has_invalid_symbols, has_double_hyphen, has_branch_name_error
```
[the remark this answers is not in the document]
Comment (NinjaWeeaboo) [reply tests/unit/branch_name_test.py]: It works, so now it's a lot cleaner thanks

File: tests/unit/branch_name_test.py [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
+
+excluded_branches = {'test-1', 'sdfsdfs', '!@dvsfosk*&*&^fs4943'}
+
+def test_excluded_branches():
```
[the remark this answers is not in the document]
Comment (NinjaWeeaboo) [reply tests/unit/branch_name_test.py]: fixed

File: tests/unit/branch_name_test.py [hunk 24]
```diff
@@ -0,0 +1,60 @@
...
+    """
+    test excluded branches checking
+    """
+    assert is_excluded_branch('test', excluded_branches) == False
```
[the remark this answers is not in the document]
Comment (NinjaWeeaboo) [reply tests/unit/branch_name_test.py]: fixed

File: tests/unit/branch_name_test.py (hunk 12 above)
Comment (NinjaWeeaboo) [reply tests/unit/branch_name_test.py]: fixed

Comment (PeterGhimself) [reply tests/unit/branch_name_test.py]: I'm nitpicking but reading that docstring kind of confuses me but "test the checking of excluded branches" sounds better

File: tests/unit/branch_name_test.py [hunk 25]
```diff
@@ -0,0 +1,63 @@
...
+    """
+    assert not is_excluded_branch('test', excluded_branches)
+    assert is_excluded_branch('test-1', excluded_branches)
+    assert is_excluded_branch('!@dvsfosk*&*&^fs4943', excluded_branches)
+
+def test_get_issue_num():
+    """
+    test get issue number function
+    """
+    assert get_issue_num('upgrade-ubuntu-18-304') == '304'
```
Comment (PeterGhimself) [inline tests/unit/branch_name_test.py]: I suggest some negative cases:
- `test`
- `18-test`
- `-18-test-735-51-&`

The function should return `None` for these cases

File: tests/unit/branch_name_test.py [hunk 26]
```diff
@@ -0,0 +1,63 @@
...
+    """
+    test invalid symbols checking
+    """
+    assert not has_invalid_symbols('feat-test-branch-123')
+    assert has_invalid_symbols('feat-test-br/anch-123')
+    assert has_invalid_symbols('feat-test-b\\ranch-123')
+    assert has_invalid_symbols('f@eat-tes?t-b\\ranch-12?3')
+    assert has_invalid_symbols('feat-test-branch-12?3')
+    assert has_invalid_symbols('!@#$-!@$%-(*&)-)()-123')
+    assert not has_invalid_symbols('----------------------18')
```
Comment (PeterGhimself) [inline tests/unit/branch_name_test.py]: What would happen if we add a negative test case `-test-branch-123`

File: tests/unit/branch_name_test.py [hunk 27]
```diff
@@ -0,0 +1,63 @@
...
+def test_has_double_hyphen():
+    """
+    test double hyphen checking
+    """
+    assert not has_double_hyphen('feat-test-branch-123')
+    assert has_double_hyphen('----------------------18')
+    assert has_double_hyphen('feat---------18')
+    assert has_double_hyphen('feat--test-branch-123')
+
+def test_has_branch_name_error():
```
Comment (PeterGhimself) [inline tests/unit/branch_name_test.py]: Maybe we should add that edge case I brought up before, where the name starts/ends with a hyphen?
like: 
- `-test-branch-name-123`
- `test-branch-name-123-`

Comment (PeterGhimself) [reply tests/unit/branch_name_test.py]: Might as well maybe add the `'!@dvsfosk*&*&^fs4943` string in here for redundancy? @MartensCedric

File: tests/unit/branch_name_test.py (hunk 12 above)
Comment (NinjaWeeaboo) [reply tests/unit/branch_name_test.py]: fixed

File: tests/unit/branch_name_test.py [hunk 28]
```diff
@@ -0,0 +1,63 @@
+import branch_name_verification_hook as branch_name_verification_hook
+from branch_name_verification_hook import *
+
+excluded_branches = {'test-1', 'sdfsdfs', '!@dvsfosk*&*&^fs4943'}
```
[the remark this answers is not in the document]
Comment (NinjaWeeaboo) [reply tests/unit/branch_name_test.py]: added

File: tests/unit/branch_name_test.py (hunk 26 above)
Comment (NinjaWeeaboo) [reply tests/unit/branch_name_test.py]: I decided to handle this case in `has_branch_name_error`. So I added a negative test case for this in `test_has_branch_name_error`

File: tests/unit/branch_name_test.py (hunk 25 above)
Comment (NinjaWeeaboo) [reply tests/unit/branch_name_test.py]: Added

File: tests/unit/branch_name_test.py (hunk 27 above)
Comment (NinjaWeeaboo) [reply tests/unit/branch_name_test.py]: Added
